README.md
[![Build Status](https://github.com/zalando-incubator/perron/workflows/CI/badge.svg?branch=master)](https://github.com/zalando-incubator/perron/actions)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# Perron

A sane client for web services with a built-in circuit-breaker, support for filtering both request and response.
```
npm install perron --save
```

**[Changelog](https://github.com/zalando-incubator/perron/blob/master/CHANGELOG.md)**

## Quick Example

The following is a minimal example of using `perron` to call a web service:

```js
const {ServiceClient} = require('perron');

// A separate instance of `perron` is required per host
const catWatch = new ServiceClient('https://catwatch.opensource.zalan.do');

catWatch.request({
    pathname: '/projects',
    query: {
        limit: 10
    }
}).then(data => console.log(data));
```

## Making Requests

Each call to `request` method will return a Promise, that would either resolve to a successful `ServiceClient.Response` containing following fields:

```js
{
    statusCode, // same as Node IncomingMessage.statusCode
    headers, // same as Node IncomingMessage.headers
    body // if JSON, then parsed body, otherwise a string
}
```

`request` method accepts an objects with all of the same properties as [https.request](https://nodejs.org/api/https.html#https_https_request_options_callback) method in Node.js, except from a `hostname` field, which is taken from the options passed when creating an instance of `ServiceClient`. Additionally you can add a `timeout` and `readTimeout` fields, which define time spans in ms for socket connection and read timeouts.

## Handling Errors

For the error case you will get a custom error type `ServiceClientError`. A custom type is useful in case you change the request to some other processing in your app and then need to distinguish between your app error and requests errors in a final catch.

`ServiceClientError` class contains an optional `response` field that is available if any response was received before there was an error.

If you have not specified retry options, you can use `instanceof` checks on the error to determine exact reason:

```js
catWatch.request({
  path: '/projects?limit=10'
}).then(console.log, logError);

function logError(err) {
  if (err instanceof BodyParseError) {
    console.log('Got a JSON response but parsing it failed');
    console.log('Raw response was', err.response);
  } else if (err instanceof RequestFilterError) {
    console.log('Request filter failed');
  } else if (err instanceof ResponseFilterError) {
    console.log('Response filter failed');
    console.log('Raw response was', err.response);
  } else if (err instanceof CircuitOpenError) {
    console.log('Circuit breaker is open');
  } else if (err instanceof RequestConnectionTimeoutError) {
    console.log('Connection timeout');
    console.log('Request options were', err.requestOptions);
  } else if (err instanceof RequestReadTimeoutError) {
    console.log('Socket read timeout');
    console.log('Request options were', err.requestOptions);
  } else if (err instanceof RequestUserTimeoutError) {
    console.log('Request dropped after timeout specified in `dropRequestAfter` option');
    console.log('Request options were', err.requestOptions);
  } else if (err instanceof RequestNetworkError) {
    console.log('Network error (socket, dns, etc.)');
    console.log('Request options were', err.requestOptions);
  } else if (err instanceof InternalError) {
    // This error should not happen during normal operations
    // and usually indicates a bug in perron or misconfiguration
    console.log('Unknown internal error');
  }
}
```

If you have retries configured, there are only 3 types of errors you will get that are relating to circuit breakers and retries, however you can access original errors that led to retries through `retryErrors` field available on both the successful response:

```js

catWatch.request({
  path: '/projects?limit=10'
}).then(function (result) {
  console.log("Response was", result.body);
  if (result.retryErrors.length) {
    console.log("Request successful, but there were retries:");
    result.retryErrors.forEach(logError);
  }
}, logError);

function logRetryError(err) {
  if (err instanceof CircuitOpenError) {
    console.log('Circuit breaker is open');
  } else if (err instanceof ShouldRetryRejectedError) {
    console.log('Provided `shouldRetry` function rejected retry attempt');
    err.retryErrors.forEach(logError);
  } else if (err instanceof MaximumRetriesReachedError) {
    console.log('Reached maximum retry count');
    err.retryErrors.forEach(logError);
  }
}
```

## Circuit Breaker

It's almost always a good idea to have a circuit breaker around your service calls, and generally one per hostname is also a good default since 5xx usually means something is wrong with the whole service and not a specific endpoint.

This is why `perron` by default includes one circuit breaker per instance. Internally `perron` uses [circuit-breaker-js](https://github.com/yammer/circuit-breaker-js), so you can use all of it's options when configuring the breaker:

```js
const {ServiceClient} = require('perron');

const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    // If the "circuitBreaker" settings are passed (non-falsy), they will be merged
    // with the default options below. Otherwise, circuit breaking will be disabled
    circuitBreaker: {
        windowDuration: 10000,
        numBuckets: 10,
        timeoutDuration: 2000,
        errorThreshold: 50,
        volumeThreshold: 10
    }
});
```

Optionally the `onCircuitOpen` and `onCircuitClose` functions can be passed to the circuitBreaker object in order to track the state of the circuit breaker via metrics or logging:

```js
const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    circuitBreaker: {
        windowDuration: 10000,
        numBuckets: 10,
        timeoutDuration: 2000,
        errorThreshold: 50,
        volumeThreshold: 10,
        onCircuitOpen: (metrics) => {
          console.log('Circuit breaker open', metrics);
        },
        onCircuitClose: (metrics) => {
          console.log('Circuit breaker closed', metrics);
        }
    }
});
```

Circuit breaker will count all errors, including the ones coming from filters, so it's generally better to do pre- and post- validation of your request outside of filter chain.

If this is not the desired behavior, or you are already using a circuit breaker, it's always possible to disable the built-in one:

```js
const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    circuitBreaker: false
});
```

In case if you want perron to still use a circuit breaker but it has to be provided dynamically by your code on-demand you can pass `circuitBreaker` option as a function (make sure to _not_ create a circuit breaker for every request):

```js
const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    circuitBreaker: function (request) {
      return somePreviouslyConstructedCB;
    }
});
```

## Retry Logic

For application critical requests it can be a good idea to retry failed requests to the responsible services.

Occasionally target server can have high latency for a short period of time, or in the case of a stack of servers, one server can be having issues
and retrying the request will allow perron to attempt to access one of the other servers that currently aren't facing issues.

By default `perron` has retry logic implemented, but configured to perform 0 retries. Internally `perron` uses [node-retry](https://github.com/tim-kos/node-retry) to handle the retry logic and configuration. All of the existing options provided by `node-retry` can be passed via configuration options through `perron`.

There is a `shouldRetry` function which can be defined in any way by the consumer and is used in the try logic to determine whether to attempt the retries or not depending on the type of error and the original request object.
If the function returns true and the number of retries hasn't been exceeded, the request can be retried.

There is also an `onRetry` function which can be defined by the user of `perron`. This function is called every time a retry request will be triggered.
It is provided the current attempt index, the error that is causing the retry and the original request params.

The first time `onRetry` gets called, the value of currentAttempt will be 2. This is because the first initial request is counted as the first attempt, and the first retry attempted will then be the second request.

```js
const {ServiceClient} = require('perron');

const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    retryOptions: {
        retries: 1,
        factor: 2,
        minTimeout: 200,
        maxTimeout: 400,
        randomize: true,
        shouldRetry(err, req) {
            return (err && err.response && err.response.statusCode >= 500);
        },
        onRetry(currentAttempt, err, req) {
            console.log('Retry attempt #' + currentAttempt + ' for ' + req.path + ' due to ' + err);
        }
    }
});
```

#### Scheduled retries

Timeouts are often the reason for retries. You might want to retry your request *without* waiting for the initial request to timeout, giving more time to the service you are calling to reply to initial request but still keeping the same overal requests timeout.

For this purpose you can use `retryAfter` in `retryOptions` in combination with `dropAllRequestsAfter` in options root. If you have retries configured and initial request has not resolved before `retryAfter`, another request will be sent and whichever requests will resolve first will be treated, aborting the other one(s).

`retryAfter` logic is immediate and will not take into account minTimeout or maxTimeout.

```js
const {ServiceClient} = require('perron');

const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    dropAllRequestsAfter: 1000
    retryOptions: {
        retries: 1,
        retryAfter: 400,
        shouldRetry(err, req) {
            return (err && err.response && err.response.statusCode >= 500);
        },
        onRetry(currentAttempt, err, req) {
            console.log('Retry attempt #' + currentAttempt + ' for ' + req.path + ' due to ' + err);
        }
    }
});
```

## Filters

It's quite often necessary to do some pre- or post-processing of the request. For this purpose `perron` implements a concept of filters, that are just an object with 2 optional methods: `request` and `response`.

By default, every instance of `perron` includes a `treat5xxAsError` filter, but you can specify which filters should be use by providing a `filters` options when constructing an instance. This options expects an array of filter object and is *not* automatically merged with the default ones, so be sure to use `concat` if you want to keep the default filters as well.

There aren't separate request and response filter chains, so given that we have filters `A`, `B` and `C` the request flow will look like this:

```
A.request ---> B.request ---> C.request ---|
                                           V
                                      HTTP Request
                                           |
A.response <-- B.response <-- C.response <--
```

If corresponding `request` or `response` method is missing in the filter, it is skipped, and the flow goes to the next one.

### Modifying the Request

Let's say that we want to inject a custom header of the request. This is really easy to do in a request filter:

```js
const {ServiceClient} = require('perron');

// A separate instance of ServiceClient is required per host
const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    filters: [{
        request(request) {
            // let's pretend to be AJAX
            request.headers['x-requested-with'] = 'XMLHttpRequest';
            return request;
        }
    }, ServiceClient.treat5xxAsError]
});
```

### Resolving Request in a Filter

Sometimes it is necessary to pretend to have called the service without actually doing it. This could be useful for caching, and is also very easy to implement:

```js
const {ServiceClient} = require('perron');

const getCache = require('./your-module-with-cache');

// A separate instance of ServiceClient is required per host
const catWatch = new ServiceClient({
    hostname: 'catwatch.opensource.zalan.do',
    filters: [{
        request(request) {
            const body = getCache(request);
            if (body) {
                const headers = {};
                const statusCode = 200;
                return new ServiceClient.Response(
                    statusCode, headers, body
                );
            }
            return request;
        }
    }, ServiceClient.treat5xxAsError]
});
```

If the request is resolved in such a way, all of the pending filter in the request and response chain will be skipped, so the flow diagram will look like this:

```
   called     |    not called
---------------------------------
 cacheFilter  |  B.treat5xxAsError
      |       |
      |       |    HTTP Request
      V       |
 cacheFilter  |  B.treat5xxAsError
```

### Rejecting Request in a Filter

It is possible to reject the request both in request and response filters by throwing, or by returning a rejected Promise. Doing so will be picked up by the circuit breaker, so this behavior should be reserved by the cases where the service returns `5xx` error, or the response is completely invalid (e.g. invalid JSON).

### JSON Parsing

By default Perron will try to parse JSON body if the `content-type` header is not set or
it is specified as `application/json`. If you wish to disable this behavior you can use
`autoParseJson: false` option when constructing `ServiceClient` object.

### UTF-8 Decoding
By default Perron will try to decode JSON body to UTF-8 string.
If you wish to disable this behaviour, you can use `autoDecodeUtf8: false` option
when calling `request` method.

### Opentracing

Perron accepts a Span like object where it will log the network and request related events.

## License

The MIT License

Copyright (c) 2016 Zalando SE

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.




CHANGELOG.md
## 0.11.5
* use ServiceClientError type for Errors in shouldRetry and onRetry options.

## 0.11.3

* Use `timing` option from client if `timing` is not set in the request options. #91
* Improved TypeScript definition of `ServiceClientRequestFilter`. #100

## 0.11.2

* Fixed an error that is thrown when `dropRequestAfter` is set and a connection error happens. #96

## 0.11.1

* Added support for measuring TLS timings.

## 0.11.0

* Added `span` option where the tcp events will be logged. The interface matches the opentracing span interface.
* Improved retries performance and memory usage.

## 0.10.0

* Added `readTimeout` options to be able to timeout when a socket is
idle for a certain period of time.

## 0.9.1

* Having a circuit breaker configured no longer results in Node process
not exiting properly.
* Improved circuit breaker performance and memory usage.

## 0.9.0

Added custom error classes for different error types, including ability to distinguish connection timeout error, user timeout error, and maximum retries error. For more details see [Handling Errors section in the README](./README.md#handling-errors)

### Breaking Changes

TypeScript type definition for request headers has been made more
strict to avoid runtime errors caused by `undefined` headers.

See [pull request](https://github.com/zalando-incubator/perron/pull/77/files) for details.

### Deprecation Notices

Usage of `type` field on `ServiceClientError` to understand the type of the error is now deprecated in favor of `instanceof` checks for new error classes added in this release.

## 0.7.0

### Breaking Changes

In 0.5.0 we changed the exports of the module to be forward-compatible with ES modules. If you are using CommonJS-style require calls, they need to updated from:

```js
const ServiceClient = require('perron')
```

to

```js
const {ServiceClient} = require('perron')
```

So `ServiceClient` is now a named export.

If you were using babel to transpile your code, no changes should be necessary.

## 0.6.0

### Breaking Changes

In 0.6.0 we changed the fields of `timings` and `timingPhases` on `ServiceClientResponse` to be nullable, or `undefined`able to be accurate. Previously `timings` had `-1` when a field was missing, and `timingPhases` had wrong numbers in those cases.


test/request.js
'use strict';

const assert = require('assert');
const sinon = require('sinon');
const proxyquire = require('proxyquire').noCallThru();
const EventEmitter = require('events');
const zlib = require('zlib');
const stream = require('stream');

class ResponseStub extends EventEmitter {}
class RequestStub extends EventEmitter {
    end() {}
}
class BufferStream extends stream.Readable {
    constructor(buffer) {
        super();
        this.i = 0;
        this.buffer = buffer;
    }
    _read() {
        const chunk = this.i < this.buffer.length ? this.buffer.slice(this.i, ++this.i) : null;
        this.push(chunk);
    }
}

describe('request', () => {

    const httpStub = {};
    const httpsStub = {};

    let request = proxyquire('../lib/request', {
        http: httpStub,
        https: httpsStub
    });
    let requestStub;

    beforeEach(() => {
        httpStub.request = sinon.stub();
        httpsStub.request = sinon.stub();
        requestStub = new RequestStub();
        httpsStub.request.returns(requestStub);
    });

    it('should call https if protocol is not specified', () => {
        request();
        assert.equal(httpsStub.request.callCount, 1);
    });

    it('should allow to call http if it is specified as protocol', () => {
        httpsStub.request.returns(undefined);
        httpStub.request.returns(requestStub);
        request({ protocol: 'http:' });
        assert.equal(httpStub.request.callCount, 1);
    });

    it('should use pathname as path if none specified', () => {
        request({pathname: '/foo'});
        assert.equal(httpsStub.request.firstCall.args[0].path, '/foo');
    });

    it('should prefer fully resolved path even if pathname is specified', () => {
        request({pathname: '/foo', path: '/bar'});
        assert.equal(httpsStub.request.firstCall.args[0].path, '/bar');
    });

    it('should allow to specify query params as an object', () => {
        request({query: {foo: 'bar', buz: 42}, pathname: '/'});
        assert.equal(httpsStub.request.firstCall.args[0].path, '/?foo=bar&buz=42');
    });

    it('should return a promise', () => {
        assert(typeof request().then, 'function');
    });

    it('should reject a promise if request errors out', (done) => {
        request().catch(() => {
            done();
        });
        requestStub.emit('error');
    });

    it('should use the body of the request if one is provided', () => {
        requestStub.write = sinon.spy();
        request({
            body: 'foobar'
        });
        assert.equal(requestStub.write.firstCall.args[0], 'foobar');
    });

    it('should resolve the promise with full response on success', (done) => {
        request().then(response => {
            assert.equal(response.body, 'foobar');
            done();
        });
        const responseStub = new ResponseStub();
        httpsStub.request.firstCall.args[1](responseStub);
        responseStub.emit('data', Buffer.from('foo'));
        responseStub.emit('data', Buffer.from('bar'));
        responseStub.emit('end');
    });

    it('should reject the promise on response error', (done) => {
        request().catch(error => {
            assert.equal(error.message, 'test');
            done();
        });
        const responseStub = new ResponseStub();
        httpsStub.request.firstCall.args[1](responseStub);
        responseStub.emit('data', Buffer.from('foo'));
        responseStub.emit('error', new Error('test'));
    });

    it('should support responses chunked between utf8 boundaries', (done) => {
        request().then(response => {
            assert.equal(response.body, 'я');
            done();
        });
        const responseStub = new ResponseStub();
        httpsStub.request.firstCall.args[1](responseStub);
        const data = Buffer.from('я');
        responseStub.emit('data', Buffer.from([data[0]]));
        responseStub.emit('data', Buffer.from([data[1]]));
        responseStub.emit('end');
    });

    ['gzip', 'deflate'].forEach((encoding) => {
        it(`should inflate response body with ${encoding} encoding`, (done) => {
            request().then(response => {
                assert.equal(response.body, 'foobar');
                assert.equal(response.statusCode, 200);
                assert.equal(response.headers['content-encoding'], 'gzip');
                done();
            }).catch(done);
            const responseStub = new BufferStream(zlib.gzipSync('foobar'));
            responseStub.statusCode = 200;
            responseStub.headers = {
                'content-encoding': 'gzip'
            };
            httpsStub.request.firstCall.args[1](responseStub);
        });
    });

    it('should reject the promise on unzip error', (done) => {
        request().catch(error => {
            assert.equal(error.message, 'incorrect header check');
            done();
        }).catch(done);
        const responseStub = new BufferStream(Buffer.from('not gzipped!'));
        responseStub.headers = {
            'content-encoding': 'gzip'
        };
        httpsStub.request.firstCall.args[1](responseStub);
    });

});


lib/client.js
'use strict';

const request = require('./request');
const url = require('url');
const CircuitBreaker = require('circuit-breaker-js');
const retry = require('retry');

/**
 * @typedef {Object.<string, (string|Array.<string>)>} ServiceClientHeaders
 */

/**
 * @typedef {{
 *   request?: ServiceClient~requestFilter,
 *   response?: ServiceClient~responseFilter
 * }} ServiceClientFilter
 */

/**
 * @typedef {{
 *  name?: string,
 *  protocol?: string,
 *  service: string,
 *  filters?: Array.<ServiceClient~requestFilter>,
 *  timing?: boolean,
 *  retryOptions?: {
 *      retries?: number,
 *      factor?: number,
 *      minTimeout?: number,
 *      maxTimeout?: number,
 *      randomize?: boolean,
 *      shouldRetry?: Function,
 *      onRetry?: Function
 *  },
 *  circuitBreaker?: (false|{
 *      windowDuration?: number,
 *      numBuckets?: number,
 *      timeoutDuration?: number,
 *      errorThreshold?: number,
 *      volumeThreshold?: number })
 *  defaultRequestOptions?: ServiceClientRequestParams
 * }} ServiceClientOptions
 */

/**
 * @typedef {{
 *  headers?: ServiceClientHeaders,
 *  method?: string,
 *  protocol?: string,
 *  port?: number,
 *  path?: string,
 *  body?: string
 *  timeout?: number,
 *  dropRequestAfter?: number,
 * }} ServiceClientRequestParams
 */

/**
 * @typedef {{
 *  headers: ServiceClientHeaders,
 *  method: string,
 *  statusCode: number,
 *  body: any,
 *  request: ?ServiceClientRequestParams
 * }} ServiceClientResponse
 */

/**
 * This callback is called after the requests has arrived
 * @callback ServiceClient~responseFilter
 * @param {ServiceClientResponse} response
 * @throws {Error}
 * @returns {ServiceClientResponse}
 */

/**
 * This callback is called before the requests is done.
 * You can short-circuit the request by both returning
 * a ServiceClient.Response object which is helpful for
 * implementing caching or mocking. You could also
 * can fail the request by throwing an Error.
 * @callback ServiceClient~requestFilter
 * @param {ServiceClientRequestParams} request
 * @throws {Error}
 * @returns {ServiceClientRequestParams|ServiceClientResponse}
 */

/**
 *
 * @param {{ headers: ServiceClientHeaders, statusCode: number, body: string, request: ServiceClientRequestParams }} response
 * @param {string} serviceClientName
 * @throws {ServiceClient.Error}
 * @returns {ServiceClient.Response}
 */
const decodeResponse = (response, serviceClientName) => {
    const contentType = response.headers['content-type'] || (response.body ? 'application/json' : '');
    if (contentType.startsWith('application/json') && typeof response.body !== 'object') {
        try {
            response.body = JSON.parse(response.body);
        } catch (error) {
            throw new ServiceClient.Error(
                serviceClientName, error, ServiceClient.BODY_PARSE_FAILED, response
            );
        }
    }
    const clientResponse = new ServiceClient.Response(
        response.statusCode,
        response.headers,
        response.body,
        response.request
    );
    clientResponse.timings = response.timings;
    clientResponse.timingPhases = response.timingPhases;
    return clientResponse;
};

/**
 * Wrapper that makes sure that all error coming out
 * of ServiceClients are of the correct type.
 * @function ServiceClient~errorWrapper
 * @param {string} type
 * @param {Error} error
 * @param {function=} responseThunk
 * @returns {Promise<Error>}
 */

/**
 * Builder for a ServiceClient~errorWrapper
 * @param {string} serviceClientName
 * @returns {ServiceClient~errorWrapper}
 */
const wrapFailedErrorBuilder = (serviceClientName) => (type, error, responseThunk) => {
    error = error instanceof ServiceClient.Error ?
        error :
        new ServiceClient.Error(serviceClientName, error, type);
    if (!error.response && responseThunk) {
        error.response = responseThunk();
    }
    return Promise.reject(error);
};

/**
 * Reducer function to unwind response filters.
 * @param {Promise.<ServiceClientResponse>} promise the promised folded upon
 * @param {ServiceClientFilter} filter the current filter in the reduce step
 * @returns {Promise.<ServiceClientResponse>}
 */
const unwindResponseFilters = (promise, filter) => {
    return promise.then(params => filter.response ? filter.response(params) : params);
};

/**
 * @param {ServiceClientRequestParams} params
 * @param {Array.<ServiceClientFilter>} filters
 * @param {string} serviceClientName
 * @returns {Promise.<ServiceClientResponse>}
 */
const requestWithFilters = (params, filters, serviceClientName) => {
    const pendingResponseFilters = [];
    const wrapFailedError = wrapFailedErrorBuilder(serviceClientName);

    const requestFilterPromise = filters.reduce((promise, filter) => {
        return promise.then(params => {
            if (params instanceof ServiceClient.Response) {
                return params;
            }
            const filtered = filter.request ? filter.request(params) : params;
            // also apply this filter when unwinding the chain
            pendingResponseFilters.unshift(filter);
            return filtered;
        });
    }, Promise.resolve(params));

    let response = null;
    let responseThunk = () => response;

    return requestFilterPromise
        .catch((err) => wrapFailedError(ServiceClient.REQUEST_FILTER_FAILED, err))
        .then((paramsOrResponse) =>
            (paramsOrResponse instanceof ServiceClient.Response) ? paramsOrResponse : request(paramsOrResponse)
        )
        .then(
            (rawResponse) => {
                response = rawResponse;
                return decodeResponse(rawResponse, serviceClientName);
            },
            (err) => wrapFailedError(ServiceClient.REQUEST_FAILED, err, responseThunk)
        )
        .then(resp => pendingResponseFilters.reduce(unwindResponseFilters, Promise.resolve(resp)))
        .catch((err) => wrapFailedError(ServiceClient.RESPONSE_FILTER_FAILED, err, responseThunk));
};

const DEFAULT_REQUEST_TIMEOUT = 2000;

/** @typedef string URL */

/**
 * @constructor
 * @property {CircuitBreaker} breaker
 * @property {ServiceClientOptions} options
 */
class ServiceClient {

    /**
     * @param {ServiceClientOptions|URL} optionsOrUrl
     */
    constructor(optionsOrUrl) {
        let options;
        if (typeof optionsOrUrl === 'string') {
            const parsed = url.parse(optionsOrUrl, true);
            const defaultRequestOptions = {};
            ['port', 'protocol', 'query', 'pathname'].forEach((option) => {
                if (parsed.hasOwnProperty(option)) {
                    defaultRequestOptions[option] = parsed[option];
                }
            });
            options = {
                hostname: parsed.hostname,
                defaultRequestOptions
            };
        } else {
            options = optionsOrUrl;
        }

        this.options = Object.assign({
            filters: ServiceClient.DEFAULT_FILTERS,
            circuitBreaker: {},
            name: options.hostname
        }, options);

        if (!this.options.hostname) {
            throw new Error('Please provide a `hostname` for this client');
        }

        this.options.defaultRequestOptions = Object.assign({
            protocol: 'https:',
            pathname: '/',
            timeout: DEFAULT_REQUEST_TIMEOUT
        }, this.options.defaultRequestOptions);

        this.options.retryOptions = Object.assign({
            retries: 0,
            factor: 2,
            minTimeout: 200,
            maxTimeout: 400,
            randomize: true,
            // eslint-disable-next-line
            shouldRetry(err, req) {
                return true;
            },
            // eslint-disable-next-line
            onRetry(currentAttempt, err, req) {}
        }, this.options.retryOptions);

        if (this.options.retryOptions.minTimeout > this.options.retryOptions.maxTimeout) {
            throw new Error('The `maxTimeout` must be equal to or greater than the `minTimeout`');
        }

        if (this.options.circuitBreaker) {
            const breakerOptions = Object.assign({
                windowDuration: 10000,
                numBuckets: 10,
                timeoutDuration: DEFAULT_REQUEST_TIMEOUT,
                errorThreshold: 50,
                volumeThreshold: 10
            }, this.options.circuitBreaker);
            this.breaker = new CircuitBreaker(breakerOptions);
        } else {
            const noop = () => {};
            this.breaker = { run(command) { command(noop, noop); } };
        }
    }

    /**
     * Perform a request to the service using given `params`.
     * @param {ServiceClientRequestParams=} userParams
     * @returns {Promise.<ServiceClientResponse>}
     */
    request(userParams) {
        const params = Object.assign({}, this.options.defaultRequestOptions, userParams);

        params.hostname = this.options.hostname;
        params.port = params.port || (params.protocol === 'https:' ? 443 : 80);

        params.headers = Object.assign({
            accept: 'application/json'
        }, params.headers);

        const {
            retries,
            factor,
            minTimeout,
            maxTimeout,
            randomize,
            shouldRetry,
            onRetry
        } = this.options.retryOptions;

        const opts = {
            retries,
            factor,
            minTimeout,
            maxTimeout,
            randomize
        };

        const operation = retry.operation(opts);

        return new Promise((resolve, reject) => operation.attempt((currentAttempt) => {
            this.breaker.run((success, failure) => {
                return requestWithFilters(params, this.options.filters, this.options.name)
                    .then(result => {
                        success();
                        resolve(result);
                    })
                    .catch(error => {
                        failure();
                        if (!shouldRetry(error, params)) {
                            reject(error);
                            return;
                        }
                        if (!operation.retry(error)) {
                            reject(error);
                            return;
                        }
                        onRetry(currentAttempt + 1, error, params);
                    });
            },
            () => {
                reject(new ServiceClient.Error(this.options.name, {}, ServiceClient.CIRCUIT_OPEN));
            });
        }));
    }
}

/**
 * @type {ServiceClientFilter}
 */
ServiceClient.treat5xxAsError = {
    response(response) {
        if (response.statusCode >= 500) {
            return Promise.reject(new Error(`Response status ${response.statusCode}`));
        }
        return response;
    }
};

/**
 * @type {ServiceClientFilter}
 */
ServiceClient.treat4xxAsError = {
    response(response) {
        if (response.statusCode >= 400 && response.statusCode < 500) {
            return Promise.reject(new Error(`Response status ${response.statusCode}`));
        }
        return response;
    }
};

ServiceClient.BODY_PARSE_FAILED = 'Parsing of the response body failed';
ServiceClient.REQUEST_FAILED = 'HTTP Request failed';
ServiceClient.REQUEST_FILTER_FAILED = 'Request filter marked request as failed';
ServiceClient.RESPONSE_FILTER_FAILED = 'Response filter marked request as failed';
ServiceClient.CIRCUIT_OPEN = 'Circuit breaker is open and prevented the request';

/**
 * Default list of post-filters which includes
 * `ServiceClient.treat5xxAsError`
 * @type {Array.<ServiceClientFilter>}
 */
ServiceClient.DEFAULT_FILTERS = Object.freeze([ServiceClient.treat5xxAsError]);

ServiceClient.Response = class ServiceClientResponse {
    // eslint-disable-next-line max-params
    constructor(statusCode, headers, body, request) {
        this.statusCode = statusCode;
        this.headers = headers;
        this.body = body;
        this.request = request;
    }
};

ServiceClient.Error = class ServiceClientError extends Error {
    // eslint-disable-next-line max-params
    constructor(name, originalError, type, response) {
        super(`${name}: ${type}. ${originalError.message || ''}`);
        Object.assign(this, originalError);
        this.type = type;
        this.response = response;
    }
};

module.exports = Object.freeze(ServiceClient);

